### PR TITLE
 Update Trade Market Structure

### PR DESCRIPTION
**Pull request by Hassan754**
Rename the Trade market structure member offer to be offer_bid, as it can be either type | refactor accordingly

### CODE REVIEW

File: integration_tests/steps/balancing_market.py
Comment (spyrostz) [inline integration_tests/steps/balancing_market.py:172]: This is a good opportunity to use the `Offer/Bid.energy_rate` member and replace the division. This applies to also other occurrences in the PR

File: integration_tests/steps/two_sided_market.py
Comment (spyrostz) [inline integration_tests/steps/two_sided_market.py:174]: Now that all these occurrences were revealed, it would be nice if we got rid of control flow via type checking. How about adding two extra methods/properties in both Offer and Bid classes, named `is_offer` and `is_bid` that will return True / False respectively and will be used for these methods ? I think it would help a lot in the readability of the caller code, if you disagree though feel free to skip it because this is purely according to personal preference.

File: src/d3a/models/market/market_structures.py
Comment (spyrostz) [inline src/d3a/models/market/market_structures.py:299]: I think you can use the classname for this, and thus removing the need for the isinstance.

File: src/d3a/models/market/market_structures.py
Comment (spyrostz) [inline src/d3a/models/market/market_structures.py:305]: You can use just `self.offer_bid.price`

File: tests/test_strategy_pvpredefined.py
Comment (spyrostz) [inline tests/test_strategy_pvpredefined.py:221]: Very minor, to replace ' with "

**Review by spyrostz — COMMENTED** [review]
Some comments, LGTM once resolved though.

**Review by hannesdiedrich — COMMENTED** [review]
LGTM once Spyros' comments are dealt with.
IMO, If there is no ticket with a specification behind the PR, there should be at least one sentence in the PR description for the rationale behind this. Even if it is only a link to a slack or confluence discussion. I kind of got the rationale behind this change from reading the diff, but a more verbose explanation than `Update Trade Market Structure` should be given.

File: integration_tests/steps/balancing_market.py
Comment (Hassan754) [reply integration_tests/steps/balancing_market.py:172]: This has exposed a bug, atm in two_sided's accept_bid we do not update the energy_rate of the bid.
I have fixed it

File: integration_tests/steps/two_sided_market.py
Comment (Hassan754) [reply integration_tests/steps/two_sided_market.py:174]: What do you think about adding these interfaces to the Trade class ? maybe named is_offer_trade() and is_bid_trade()

Comment (spyrostz) [reply integration_tests/steps/two_sided_market.py:174]: Even better! As long as the check is self-contained in the Trade class and not done by the caller code it works for me, thanks!

**Review by hannesdiedrich — DISMISSED** [review]
LGTM

File: src/d3a/models/market/market_structures.py
Comment (fievelk) [inline src/d3a/models/market/market_structures.py:78]: I would personally duplicate this code and put it both in Offer and Bid. I think it would be a bit more appropriate than having this `if` here.

File: src/d3a/models/market/market_structures.py
Comment (fievelk) [inline src/d3a/models/market/market_structures.py:103]: Nit: since we're unifying the interface for Bids and Offers, I got to think that these are basically the same concepts:
- `self.seller` and `self.buyer`
- `self.seller_origin` and `self.buyer_origin`

So they could in theory be unified and put inside `BidOffer`. The difficult part is the naming, though. What do these represent for Bid/Offer objects? Some hypotheses:
- `self.seller` and `self.buyer` --> `self.parent_origin` or `parent_actor` or `parent_trader`
- `self.seller_origin` and `self.buyer_origin` --> `self.root_origin` or `root_actor` or `root_trader`

Maybe you can come up with something better. If you think that this only complicates things, no need to do anything.

File: src/d3a/models/market/market_structures.py
Comment (fievelk) [inline src/d3a/models/market/market_structures.py:143]: What about renaming this method so that it will be the same for Offer and Bids? E.g.
```suggestion
def copy(offer) -> Offer:
# or
def create_duplicate(offer) -> Offer:
```

File: src/d3a/models/market/market_structures.py
Comment (fievelk) [inline src/d3a/models/market/market_structures.py:102]: As I suggested on slack, this could also be renamed to be 
```suggestion
        self.original_price = original_offer_price
```
and put into `BidsOffer`

File: src/d3a/models/market/market_structures.py
Comment (fievelk) [inline src/d3a/models/market/market_structures.py:206]: I think that this would be a good candidate to be a class method for `BaseBidOffer`:
```python
BaseBidOffer.from_json()
```
(Also, no need to specify the `_string` part in `from_json_string`, I think).

File: src/d3a/models/market/market_structures.py
Comment (fievelk) [inline src/d3a/models/market/market_structures.py:213]: Nit: I have the feeling that this logic doesn't belong here. But I see that you just took it from the existing version, so no need to fix it here.

File: src/d3a/models/market/market_structures.py
Comment (fievelk) [inline src/d3a/models/market/market_structures.py:231]: Nit: where is this used?

File: src/d3a/models/market/market_structures.py
Comment (fievelk) [inline src/d3a/models/market/market_structures.py:235]: Similar comment as above: I would see this as a class method for `TradeBidOfferInfo`:
```python
trade_bid_offer = TradeBidOfferInfo.from_json(json_str)
```

Comment (fievelk) [reply src/d3a/models/market/market_structures.py:235]: Nit: On another side: why is this called `TradeBidOfferInfo` and not `TradeInfo`? Do you have any idea about the reason?

File: src/d3a/models/market/market_structures.py
Comment (fievelk) [inline src/d3a/models/market/market_structures.py:271]: Nit: two points about this method:
1. the method name is a bit misleading. The output is not a CSV file, but rather a tuple that can be used for many purposes (one of those is to create a CSV file);
2. the method should be public, since it's being used outside of this class

I know this is not part of the PR, but I'm taking this opportunity that you gave us with the refactoring to note it.

File: src/d3a/models/market/market_structures.py
Comment (fievelk) [inline src/d3a/models/market/market_structures.py:289]: Nit: a possible alternative would just be
```suggestion
    def is_bid(self) -> bool:
```
Do you think this would be accurate? (Same for the offer part)

File: src/d3a/models/strategy/__init__.py
Comment (fievelk) [inline src/d3a/models/strategy/__init__.py:461]: Could you please add a type hint to the `trade` argument of this method? I suppose it is a `TradeBidOfferInfo` and not a `Trade` instance, right?

**Review by fievelk — COMMENTED** [review]
Left some comments. Thanks for the refactoring!

**Comment by Hassan754** [comment]
@fievelk Thanks for your comments, I agree with the majority of them
In order to not overcomplicate this PR and make the review process harder, I'm thinking about creating another PR to address them
In this case, do you think it makes sense to gradually introduce these changes into the master branch (Merging the changes introduced here then the ones in the next PR)?

**Comment by fievelk** [comment]
> In order to not overcomplicate this PR and make the review process harder, I'm thinking about creating another PR to address them

@Hassan754 Since my comments were not major nor blocking, that's definitely fine for me.

**Comment by Hassan754** [comment]
> > In order to not overcomplicate this PR and make the review process harder, I'm thinking about creating another PR to address them
> 
> @Hassan754 Since my comments were not major nor blocking, that's definitely fine for me.

Awesome, I'll move forward with this PR and keep it as a reference to address the comments

**Review by fievelk — APPROVED** [review]

**Comment by fievelk** [comment]
@Hassan754 regarding your latest commit: https://github.com/gridsingularity/d3a/pull/1184/commits/10120fdf51268c6e140476ed9b490bba1bd0506d

This bug should have made tests fail. Since it didn't happen, I suppose that this part of code is not tested. It would be great if you could add tests for it and make sure they would have failed with the previous bug.

**Comment by Hassan754** [comment]
> @Hassan754 regarding your latest commit: [10120fd](https://github.com/gridsingularity/d3a/commit/10120fdf51268c6e140476ed9b490bba1bd0506d)
> 
> This bug should have made tests fail. Since it didn't happen, I suppose that this part of code is not tested. It would be great if you could add tests for it and make sure they would have failed with the previous bug.

Actually they were failing, there is a dedicated unit test function that was testing this, it failed after resolving the merge conflicts with master

File: src/d3a/models/market/market_structures.py
Comment (hannesdiedrich) [reply src/d3a/models/market/market_structures.py:289]: No, IMO this would be confusing. A trade can not be a bid. This is to check if the trade was issued when accepting a bid. Please leave the property name as is, thanks!

**Review by hannesdiedrich — APPROVED** [review]
LGTM

**Pull request merged**
